### PR TITLE
chore: hoist the TTL bounds behind one canonical constant

### LINKED ISSUE #315: Agree on canonical TTL range and move 10..86400 into a shared constant
## Summary

The TTL bounds `10..86400` (10 seconds → 24 hours) are currently **hardcoded in multiple places** and were never explicitly agreed on as a product contract. We should (a) decide whether this range is right, and (b) move it into a single shared constant that all packages import, instead of re-typing the literals.

## Where the range lives today

The `10..86400` bound originated in the **agent** proposal schema (PR #134, `aa364add`, @jamby77):

- `packages/shared/src/utils/cache-proposals.ts:47`
  ```ts
  new_ttl_seconds: z.number().int().min(10).max(86400),
  ```

PR #161 (semantic-cache runtime TTL) then **copied the same literals** into the library layer, in two more places:

- `packages/semantic-cache/src/SemanticCache.ts` — `refreshConfig()` validates `parsed >= 10 && parsed <= 86400`
- `packages/semantic-cache-py/betterdb_semantic_cache/semantic_cache.py` — `refresh_config()` validates `10 <= parsed <= 86400`

So the same magic numbers now exist as **three independent copies** (one Zod schema + a TS literal + a Python literal) with nothing tying them together.

## Problems

1. **The range may be too limiting.** A 24-hour ceiling is arbitrary. Legitimate use cases (e.g. a semantic cache constructed with `defaultTtl: 604800` / 7 days) are honored at the constructor level but **cannot be set via the runtime `__config` override or a TTL proposal**, because anything above 86400 is silently rejected. Likewise the 10s floor is a guess. We should agree on values deliberately (or make them configurable).

2. **Silent divergence risk.** Because the numbers are duplicated, a future `SemanticTtlAdjustPayloadSchema` (the apply-side of #161) could easily use a different range than the library enforces, so the API would accept a value the library rejects, or vice versa. This was flagged by @amitkojha05 in #161 and is currently unresolved.

3. **Cross-language duplication.** TS and Python each carry their own copy, so any change requires editing both and keeping them in sync by hand.

## Proposed direction (to discuss)

- Decide the canonical `[MIN_TTL_SECONDS, MAX_TTL_SECONDS]` range (and whether it should be configurable rather than fixed).
- Lift it into a single shared constant:
  - TS: export from `packages/shared/src/utils/cache-proposals.ts` (e.g. `TTL_SECONDS_MIN` / `TTL_SECONDS_MAX`) and have the Zod schema **and** `@betterdb/semantic-cache` / `@betterdb/agent-cache` import it.
  - Python: a matching module-level constant that `semantic_cache.py` (and the py agent-cache) import, kept in sync with the TS source of truth.
- Update the schema + both library validators to reference the constant instead of literals.

## Related

- Origin of the bound: PR #134 (`aa364add`)
- Duplication introduced by: PR #161 (semantic-cache runtime TTL)
- Owner context: @jamby77 set the original `10..86400`; @amitkojha05 raised the divergence concern

cc @jamby77 @amitkojha05

### PR DESCRIPTION
**Pull request by jamby77**
Closes #315.

The proposal TTL bound `10..86400` was written out as a literal in five places. This hoists it behind `TTL_SECONDS_MIN` / `TTL_SECONDS_MAX` and makes drift a CI failure rather than a convention.

No behaviour change — every declaration still resolves to `10` and `86400`.

## Why mirrors rather than a shared import

`@betterdb/shared` is `private: true` and is never published. `@betterdb/semantic-cache`, `@betterdb/mcp` and the `betterdb-semantic-cache` wheel are each installed standalone by external users, so none of them can import it without breaking `npm install` / `pip install`. Each keeps a local mirror instead, and CI enforces that the mirrors agree with the canonical.

This is @amitkojha05's proposal from the issue, with one amendment: he wrote that cross-artifact sync stays by convention. It doesn't have to — CI scripts are dev-only, so the monorepo can verify every mirror without any published artifact gaining a runtime dependency.

## The five sites

| File | Was |
|---|---|
| `packages/shared/src/utils/cache-proposals.ts` | `z.number().int().min(10).max(86400)` |
| `packages/semantic-cache/src/SemanticCache.ts` | `parsed >= 10 && parsed <= 86400` |
| `packages/semantic-cache-py/betterdb_semantic_cache/semantic_cache.py` | `if 10 <= parsed <= 86400` |
| `packages/mcp/src/index.ts` (`cache_propose_tool_ttl_adjust`) | `.min(10).max(86400)` |
| `packages/mcp/src/index.ts` (`cache_edit_and_approve_proposal`) | `.min(10).max(86400)` |

The issue named three; there are five. `agent-cache` (TS and py) has no TTL validator of its own, so nothing changed there despite being listed.

The `cache_propose_tool_ttl_adjust` description string (`'Proposed TTL in seconds (10–86400)'`) is now a template literal, so the documented range cannot drift from the enforced one.

Unrelated `86400`s — the embedding-cache TTL default, formatters, seed data, the langgraph examples — were left alone.

## The guard

`scripts/check-ttl-constants.mjs` reads the declared values **as source text** and fails if they disagree. Reading text rather than importing is what lets one check cover TypeScript and Python alike, with no build, no install and no test runner in any package.

That last part matters: `api-tests.yml` is the only other `pull_request`-triggered workflow, and it does not run `packages/semantic-cache`'s vitest suite. `@betterdb/mcp` has no test runner at all. A drift test living in either package would gate nothing, which is worse than no guard because it reads as protection. The guard needs neither.

The mirror set is not hand-maintained. The script walks `apps`, `packages`, `proprietary` and `scripts` and checks every declaration of `TTL_SECONDS_MIN` / `TTL_SECONDS_MAX` it finds, so a mirror added anywhere in the repo — including outside `packages/**` — is covered the moment it exists. It also fails on a bare `10`/`86400` written as a two-sided schema bound (`.min(10).max(86400)`, `Field(ge=10, le=86400)`) or a two-sided clamp, which is the drift that matters: reverting an MCP schema to `.max(9999)` is a CI failure, not a review catch.

It also catches the case the mirrors alone miss: editing the canonical and forgetting the mirrors.

**Triggering.** The guard lives in its own workflow, `.github/workflows/constants-guard.yml`, on `pull_request` to `master` with no `paths:` filter — a path filter would let a mirror added outside the filtered tree escape the trigger, which defeats the point. This branch does not modify `api-tests.yml`.

I confirmed it fails rather than assuming it does. Mutation-tested across seven scenarios: mismatched mirror, half-declared pair, annotated declaration, commented declaration, literal schema bound, literal clamp, and a one-sided bound that must *not* be reported.

## Tests

The `config-refresh` boundary cases in both languages now derive their inputs from the constants instead of hardcoding `9` and `86401`, and each gains an accepts-at-the-boundary case.

The reject-side test is the one that bites: set a mirror's min to `11` while leaving the validator hardcoded at `10` and `test_ttl_ignores_below_minimum` fails. The accept-side test cannot detect drift on its own, since it feeds the same constant it asserts on — it is there to document the boundary.

The Python suite is PR-gated, so that half is real coverage today. The TypeScript half becomes coverage the moment that suite is wired into CI.

## Not done here

The constants are not re-exported from either package's public entry point — smaller diff, no new public API surface, and it avoids touching the `@betterdb/ai` completeness guard. Easy to add if reviewers want the contract inspectable by callers.

Whether the ceiling should move or become configurable is deliberately out of scope. Item (1) of #315 describes a real inconsistency — a `defaultTtl: 604800` cache is legal at construction but cannot be set through the runtime `__config` override or a proposal — and that deserves its own PR reviewed on its own merits. After this lands it is a one-line change per artifact, and the guard makes a partial change impossible to merge.

## Verification

- `node scripts/check-ttl-constants.mjs` — passes across 4 declarations; verified it fails on drift and on a missing declaration
- `@betterdb/shared` and `@betterdb/mcp` build
- `packages/semantic-cache`: `tsc --noEmit` clean, 229 vitest pass
- `packages/semantic-cache-py`: 191 pytest pass
- `@betterdb/ai` completeness guard: 23 pass

`apps/api` jest has 25 pre-existing failures on `master` at 1e812c32, unrelated to this change and unchanged by it (verified by stashing and rerunning the same suites).

---

> [!NOTE]
> **Low Risk**
> Refactor plus CI enforcement only; TTL limits are unchanged and drift is caught before merge.
> 
> **Overview**
> Replaces scattered **10–86400** TTL literals with shared **`TTL_SECONDS_MIN`** / **`TTL_SECONDS_MAX`**, with **`packages/shared/src/utils/cache-proposals.ts`** as the canonical definition and mirrored copies in **`@betterdb/mcp`**, **`@betterdb/semantic-cache`**, and the Python semantic-cache wheel (published packages cannot depend on private **`@betterdb/shared`**).
> 
> Validation and runtime config refresh now reference those names in proposal schemas, MCP TTL tools, and semantic-cache **`__config`** TTL parsing. **No intended behavior change** — bounds stay 10s and 86400s.
> 
> Adds **`scripts/check-ttl-constants.mjs`** and a **`Constants Guard`** GitHub Actions workflow on PRs to **`master`**: it compares every discovered mirror to the canonical file and fails on literal min/max in TTL schema or clamp sites. Config-refresh tests in TS and Python derive edge cases from the constants and add explicit min/max acceptance cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 705bb9b54270bb3963ec0758bdada7588a1827bf. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Summary by CodeRabbit

* **Improvements**
  * Standardized cache TTL limits across supported integrations and configuration tools.
  * TTL values now consistently accept 10–86,400 seconds and reject values outside that range.

* **Bug Fixes**
  * Prevented inconsistent TTL validation between components.

* **Tests**
  * Added coverage for minimum and maximum TTL boundaries.

* **Chores**
  * Added automated checks to detect future TTL limit drift.

### CODE REVIEW

File: .github/workflows/api-tests.yml [hunk 1]
```diff
@@ -16,8 +16,25 @@ on:
...
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.branch || github.ref }}
+
+      - uses: actions/setup-node@v4
+        with:
+          node-version: '20'
+
+      - name: Check canonical TTL bounds against every mirror
+        run: node scripts/check-ttl-constants.mjs
```
[the remark this answers is not in the document]
Comment (jamby77) [reply .github/workflows/api-tests.yml]: Fixed in c0fe64fc — `permissions: contents: read` on `constants-guard`.

File: .github/workflows/api-tests.yml [hunk 2]
```diff
@@ -16,8 +16,25 @@ on:
...
 
 jobs:
+  constants-guard:
+    runs-on: ubuntu-latest
+    timeout-minutes: 5
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ inputs.branch || github.ref }}
```
[the remark this answers is not in the document]
Comment (jamby77) [reply .github/workflows/api-tests.yml]: Fixed in c0fe64fc — `persist-credentials: false` on the `constants-guard` checkout. The pre-existing checkouts in this workflow are untouched; hardening those is out of scope for a TTL-constant PR.

[the remark this answers is not in the document]
Comment (jamby77) [reply .github/workflows/api-tests.yml]: @coderabbitai no issue needed — the repo-wide `persist-credentials` pass is tracked internally, along with the current counts (24 unhardened checkout steps across 17 workflow files, 6 already hardened) and the finding that no workflow relies on the persisted token, so every one of them can be flipped. It will be sequenced with the action SHA-pinning pass, since both touch the same files.

**Review by KIvanow — CHANGES_REQUESTED** [review]
The refactor itself is correct and behaviour-preserving, and the transitive server-side validation is the right pattern. But I'd like three things addressed before this lands, since the whole value proposition is "drift becomes a CI failure rather than a convention" and right now the guard doesn't fully deliver that.

1. Fix the guard's brittleness (check-ttl-constants.mjs). The Python regex ^NAME\s*=\s*(\d+)\s*$ doesn't match a type annotation (TTL_SECONDS_MIN: int = 10) or a trailing # comment - both normal edits. It fails safe (CI goes red), but it'll red CI on perfectly good code with a misleading "no declaration found." Same for the TS side with a trailing // comment. Please loosen both patterns to tolerate annotations and inline comments.

2. Close the usage gap. The guard keeps the three mirror declarations numerically equal, but nothing stops a bare 10/86400 from reappearing at a call site - which is the drift we actually care about. As written, someone can revert an MCP schema to .max(9999) and the guard passes green. Please add a lint rule (or an equivalent check) banning the bare literals at the consumer sites so the constant is the only legal way to express the bound. That's the guarantee the PR is selling.

3. Make coverage self-enforcing. The MIRRORS list is hand-maintained with nothing enforcing completeness, and the workflow's path filter only really covers packages/** - a mirror added outside that (a top-level SDK, a docs example) silently escapes both the list and the trigger. Please either derive the mirror set rather than hardcode it, or add a check that fails when a declared constant exists outside the known set. And correct the PR description: the job triggers via packages/**, not scripts/** as written.

File: .github/workflows/api-tests.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jamby77) [reply .github/workflows/api-tests.yml]: Moot as of 210d6b2d: this branch no longer touches `.github/workflows/api-tests.yml` at all (it is back to its master state here). The TTL guard moved to its own `.github/workflows/constants-guard.yml`, which has no path filter — a new mirror of the constants cannot be added in a PR that skips the trigger — and that workflow does set `persist-credentials: false` on its checkout.

**Comment by jamby77** [comment]
Addressed in 210d6b2d. `scripts/check-ttl-constants.mjs` was rewritten rather than patched, because items 2 and 3 changed what the script has to be.

**1. Declaration patterns.** Both sides now tolerate a type annotation and a trailing comment:

- TS: `^export const NAME\s*(?::[^=\n]+)?=\s*(\d+)\s*;?\s*(?://.*)?$`
- Py: `^NAME\s*(?::[^=\n]+)?=\s*(\d+)\s*(?:#.*)?$`

`TTL_SECONDS_MIN: int = 10  # seconds` and `export const TTL_SECONDS_MAX: number = 86400; // one day` both parse. A file that declares one of the pair without the other is reported by name instead of silently half-checked.

**2. Usage gap.** The guard now fails on a bare literal at a call site, in the two shapes that actually state the range:

- a two-sided schema bound — `z…min(10).max(86400)`, `Field(ge=10, le=86400)`. A one-sided bound is not the canonical range and is left alone, which is what keeps `ANOMALY_CACHE_TTL_MS.min(1000)` and `current_ttl_seconds.min(0)` out of the report.
- a two-sided clamp — `ttl >= 10 && ttl <= MAX`, `10 <= ttl <= MAX`, and the reversed spellings.

Verified against your exact regression: reverting an MCP schema to `.max(9999)` is caught, and so is `.max(86400)` written as a literal.

**3. Self-enforcing coverage.** `MIRRORS` is gone. The script walks `apps`, `packages`, `proprietary` and `scripts` for `.ts/.tsx/.py/.mjs/.js`, finds every declaration of the two names wherever it lives, and compares them all against the canonical file — so a mirror in a top-level SDK or a docs example is checked the moment it exists, and cannot escape by not being on a list. Current run:

```
TTL constant guard passed: TTL_SECONDS_MIN=10 TTL_SECONDS_MAX=86400 across
4 declarations, no literal bounds in 1425 source files.
```

The path filter is gone too. The guard moved out of `api-tests.yml` into its own `.github/workflows/constants-guard.yml`, which triggers on every PR to master with no `paths:` — a mirror added outside `packages/**` can no longer escape the trigger. That also means this branch no longer touches `api-tests.yml` at all.

Mutation-tested across seven scenarios (mismatched mirror, half-declared pair, annotated declaration, commented declaration, literal schema bound, literal clamp, one-sided bound) before pushing.

PR description corrected.

File: scripts/check-ttl-constants.mjs [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+
+/**
+ * A declaration line is where the literal is supposed to live, so it is not
+ * itself a call site.
+ */
+function isDeclarationLine(line) {
+  return NAMES.some((name) => {
+    return line.includes(name);
+  });
+}
```
[the remark this answers is not in the document]
Comment (jamby77) [reply scripts/check-ttl-constants.mjs:202]: Confirmed and fixed in `f7157de6`. Reproduced it against the exact code you point at — `SemanticCache.ts:1257` — by rewriting the clamp to `parsed >= TTL_SECONDS_MIN && parsed <= 86400`:

```
$ node scripts/check-ttl-constants.mjs      # before
TTL constant guard passed: TTL_SECONDS_MIN=10 TTL_SECONDS_MAX=86400 across 4 declarations…

$ node scripts/check-ttl-constants.mjs      # after
  - packages/semantic-cache/src/SemanticCache.ts:1257: a TTL range check uses the literal 86400
```

A half-drifted ceiling merged clean. `isDeclarationLine` now requires an actual assignment rather than a mention:

```js
return new RegExp(`^\\s*(?:export\\s+const\\s+)?${name}\\s*(?::[^=\\n]+)?=`).test(line);
```

Both orientations are covered — constant floor with a literal ceiling, and literal floor with a constant ceiling.

File: scripts/check-ttl-constants.mjs [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+ * actually state the bound, and only those are guarded:
+ *
+ *   1. A schema field named for a TTL carrying a `.min()`/`.max()` or a
+ *      pydantic `ge=`/`le=` bound. This is the `.max(9999)` regression.
+ *   2. A two-sided range in a stretch of code that is handling a TTL, which is
+ *      the runtime clamp. One-sided comparisons are left alone.
+ */
+
+/** A schema field declaring a TTL: `new_ttl_seconds: z…` or `ttl: Field(…)`. */
+const TTL_SCHEMA_FIELD = /^[ \t]*([A-Za-z_]\w*ttl\w*)\s*:\s*(?:z\b|\w*Field\()/gim;
```
[the remark this answers is not in the document]
Comment (jamby77) [reply scripts/check-ttl-constants.mjs]: Right about the annotated form, and the gap ran deeper than the annotation. `([A-Za-z_]\w*ttl\w*)` requires at least one character *before* `ttl`, so a field simply called `ttl_seconds` or `ttl` never matched — annotated or not. Only `new_ttl_seconds` and friends were ever checked.

Fixed in `f7157de6` by not encoding the name test in the pattern at all:

```js
const SCHEMA_FIELD = /^[ \t]*([A-Za-z_]\w*)\s*:\s*(?:z\b|(?:[^=\n]+=\s*)?\w*Field\()/gim;
const TTL_FIELD_NAME = /ttl/i;
```

with the name filtered in the loop. The optional `(?:[^=\n]+=\s*)?` is your annotation fix; `NEXT_SCHEMA_FIELD` got the same clause so an annotated sibling still terminates the span.

Mutation-tested across seven schema-field scenarios: annotated pydantic with literal bounds fires, annotated with named bounds does not, bare `Field(` fires, a non-TTL `port: int = Field(6379, ge=1, le=65535)` does not, a zod TTL field reverted to `.max(9999)` fires, the same field with named bounds does not, and a one-sided `.min(0)` does not.

File: scripts/check-ttl-constants.mjs [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+  for (const { pattern } of RANGE_BOUNDS) {
+    pattern.lastIndex = 0;
+    for (const hit of source.matchAll(pattern)) {
+      const literal = hit[2] ?? hit[1];
+      if (/^\d+$/.test(literal) === false) {
+        continue;
+      }
+
+      const before = source.slice(Math.max(0, hit.index - CLAMP_WINDOW), hit.index);
+      if (TTL_MENTION.test(before) === false) {
```
[the remark this answers is not in the document]
Comment (jamby77) [reply scripts/check-ttl-constants.mjs]: Both halves valid, both fixed in `f7157de6`, and I took your two diffs essentially as written.

The `before`-window half: a clamp whose only TTL mention is inside the match reads as an unrelated comparison. Now tested against the range text too.

```js
if (TTL_MENTION.test(`${before}\n${hit[0]}`) === false) {
```

The `isDeclarationLine` half is the same bug Bugbot filed from the other side, and is fixed with your assignment-anchored form. It reproduced against `SemanticCache.ts:1257` — a clamp keeping `TTL_SECONDS_MIN` and hard-coding `86400` passed the old guard.

One thing the fix surfaced: with the wider mention test, the guard failed on its own source, because the comment explaining the change contains `ttl >= 10 && ttl <= TTL_SECONDS_MAX` as prose. A range in a comment describes a bound rather than applying one, so comment lines are now skipped alongside declaration lines. Only the line the match starts on is tested, so code with a trailing comment is still a call site.

Verified: `ttl >= 10 && ttl <= 86400` with no preceding TTL mention now fires; `port >= 1 && port <= 65535` still does not.

**Review by KIvanow — APPROVED** [review]
